Design for a "files/objects" schema

brianaydemir opened the pull request:
As part of the CriticalMAAS project, we need to design and implement a pipeline for ingesting maps into Macrostrat. This pull request is for a design document that covers how we track "files" that need to be processed/ingested.

brianaydemir commented:
Feedback from @davenquinn (copied from Slack):

Thanks for creating this! Few potential thoughts:

* I think there should be another table that records bucket-level metadata, which might contain S3 role information and other attributes

* I agree with [@Cannon Lock](https://uw-xdd.slack.com/team/U05FQKTDSQ4) that there should be some way of easily constructing a "naïve" S3 url that requires the calling script to manage the auth appropriately. In many cases , it will be more efficient/performant for our internal scripts to handle that up front for buckets we know about

* This S3 url construction could be via a SQL function, if we didn't want to store it directly

* Along with the S3 url consideration, do we want to consider/handle the edge case of files that are accessible via http/s but not in an S3 bucket specifically? I'm thinking about USGS object storage/cloudfront, which provides https urls that are very S3-like (i.e. could be referenced in place), but obfuscates the internal details of the specific storage system

* We should probably have several optional fields for file management and tracking, that will ease use in many cases. I'm thinking something like:

  * mime_type
  * file_hash
  * date_created
  * date_last_accessed
  * date_deleted
 
  The date_deleted field might not be the right way to handle this, but I am essentially thinking about a way to "mark a file for deletion" that can allow a later batch script to remove or "forget" the file as appropriate for a specific bucket.

davenquinn commented:
Now that I'm looking at requirements here, the "supabase storage schema" that I have mentioned a few times looks pretty similar to the design we're developing here:

https://github.com/UW-Macrostrat/stratigram/blob/main/schema/04-storage-system.sql: Tables and helpers for file search, path management, etc.

The big difference from our needs, i think, is that for Supabase, the "bucket" is actually an internal concept — all files are stored in the same S3 bucket actually, there is just an additional "file grouping" concept that mimics buckets or folders for organization and app-level auth purposes. We'll probably want the option to work across multiple S3 buckets for our files, but this could be implemented as another key in the buckets table, for instance.

brianaydemir commented:
I've updated the design.

* I tried to generalize it to more than just Amazon S3-compatible object stores.

* I added additional columns for the requested metadata.

* I included the "naïve URL" as an implementation consideration. I don't see these URLs as a proper component of the data model (there's no reason to assume that that's how one accesses an object store). And for something like authenticated S3 requests, the (few) libraries I'm aware of are going to make you specify the host, bucket, and key separately anyway.

CannonLock reviewed (approved):
LGTM

davenquinn commented on docs/Map Ingestion - Files Schema.md:
```diff
@@ -0,0 +1,88 @@
+
+- id: Primary key. An integer with no other meaning attached to it.
+
+- scheme: A string indicating the kind of object store. Examples include:
```
I think I like this, but do you think it is too much of a mess to store random `http` URLs? It seems like there would need to be some sort of guarantee that the file is "s3-object-like" in some way...

Overall, I think this is good in that it preserves the option to track a variety of things in the future.

davenquinn commented on docs/Map Ingestion - Files Schema.md:
```diff
@@ -0,0 +1,88 @@
+
+- key: Not null.
+
+- (scheme, host, bucket, key): Unique.
```
I wonder if the `scheme`, `host`, and `bucket` should go in a separate "bucket" table? This would allow easier bulk transfers of objects, access control changes, etc. and be slightly more normalized. Of course, inserts would be more complex.

This would also be an awkward fit with the `http` scheme.

davenquinn replied:
This would presumably where bucket credentials would be stored, if applicable.

davenquinn replied:
Relatedly, should we define at this point how groupings (application-level, not object-store-level) might work? Will there be a single approach or will it be up to whatever maintains foreign keys into this file schema?

I can see advantages of standardizing this and requiring that all objects added are assigned to a specific group. But it might be more complexity for little benefit.

brianaydemir replied:
There is a pathological case where each individual object might have its own access credentials and policy. That said, it becomes a management nightmare if you don't set that all that bucket-level.

Regardless, there should probably be a second table that tracks access credentials and whatnot, and way to link it to the "objects" table.

davenquinn commented on docs/Map Ingestion - Files Schema.md:
```diff
@@ -0,0 +1,88 @@
+- key: A string containing the object's name within the object store and,
+  if applicable, bucket.
+
+- source: JSON object describing the provenance of the object indicated by
```
Can you give an example of what might be in this JSON object?

davenquinn commented on docs/Map Ingestion - Files Schema.md:
```diff
@@ -0,0 +1,88 @@
+For objects that can be retrieved using a simple, unauthenticated HTTP GET
+request, it may be helpful to have a computed or virtual column containing
+the requisite URL.
+
```
Planning ahead, it might also be useful to suggest the direction of travel for how access control to specific objects could work (I'm hoping that we can use Postgres **Row Level Security** for this)

davenquinn replied:
Another meta-question: without outgoing links to other tables that use the objects, or predefined groupings, will we have any clear method to evaluate when objects are unused by other parts of the schema?

I think this is probably out of scope but it's worth considering at some level, probably.

CannonLock replied:
@brianaydemir @davenquinn I was imagining the grouping was done by zipping the files together and entering them as one object. If that is not the case or we want to start with something more flexible it would be great to have a group_id that signifies a group of files.

davenquinn replied:
@CannonLock I agree that makes sense for the case of storing files for map ingestion. I was thinking more about tracking broader groups of objects for specific application-level purposes, such as **Maps ingested from USGS CloudFront** or **Maps dumped to TA1 output schema" or "Raster maps stored at Jataware". Perhaps we'd need to easily surface, filter by, or manage such sets as a whole.

CannonLock replied:
@davenquinn How about a FK for the user that added the map? That user will have all groups that it is related too attached.

brianaydemir replied:
> Planning ahead, it might also be useful to suggest the direction of travel for how access control to specific objects could work (I'm hoping that we can use Postgres Row Level Security for this).

Are we talking about access to knowledge that the object exists in Macrostrat? Or access to the object itself? Those are two different things, and a priori, not necessarily aligned.

For example, for anything in S3, we can hand the end-user a time-limited, pre-signed URL for downloading the object (or uploading a new object). At no point do we need to give them actual S3 access and secret keys. We can use whatever machinery we want for determining whether we _should_ give them such a pre-signed URL in the first place.

(Row level security strikes me as odd unless we're making a DB user for each end-user? Maybe I'm not understanding something.)

davenquinn reviewed (commented):
I like the addition for scheme (`s3` vs. `http`). That's potentially very useful.

Overall I think this basic design is quite reasonable; I do wonder if there should be another table to normalize information about buckets and file groups. We could also punt on this.

Perhaps these aspects could be added to a new section **Potential Future Enhancements** if we want to just build the basic thing without trying to handle all potential elaborations.

I left several comments along those lines and

CannonLock commented:
@davenquinn @brianaydemir I am going to grab this state as it is now to get things started.

davenquinn commented:
@brianaydemir @CannonLock, aligned with #25 can we merge this if appropriate or mark it as superseded (and by what?)
